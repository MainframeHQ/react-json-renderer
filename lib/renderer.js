'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Renderer = exports.renderFromObject = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Fallback = function Fallback() {
  return null;
};

var renderFromObject = function renderFromObject(tree, params) {
  var components = params && params.components || {};
  var fallback = params && params.fallback || Fallback;

  var createChild = function createChild(c) {
    if (c == null) {
      return null;
    }
    if (typeof c === 'string') {
      return c;
    }
    if (c.type) {
      return createComponent(c);
    }
    return null;
  };

  var createComponent = function createComponent(converted) {
    var component = components[converted.type] || fallback;

    var _converted$props = converted.props,
        children = _converted$props.children,
        props = _objectWithoutProperties(_converted$props, ['children']);

    if (children) {
      props.children = Array.isArray(children) // eslint-disable-line react/prop-types
      ? children.map(createChild) : createChild(children);
    }
    return (0, _react.createElement)(component, props);
  };

  return createComponent(tree);
};

exports.renderFromObject = renderFromObject;

var Renderer = exports.Renderer = function (_Component) {
  _inherits(Renderer, _Component);

  function Renderer() {
    _classCallCheck(this, Renderer);

    return _possibleConstructorReturn(this, (Renderer.__proto__ || Object.getPrototypeOf(Renderer)).apply(this, arguments));
  }

  _createClass(Renderer, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          json = _props.json,
          tree = _props.tree,
          params = _objectWithoutProperties(_props, ['json', 'tree']);

      var obj = void 0;
      if (tree) obj = tree;else if (json) obj = JSON.parse(json);

      return obj ? renderFromObject(obj, params) : null;
    }
  }]);

  return Renderer;
}(_react.Component);