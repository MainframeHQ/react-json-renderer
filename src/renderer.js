// @flow

import React, { createElement, PureComponent, type ComponentType } from 'react'

import type { ConvertedElement } from './types'

type RenderParams = {
  components?: { [type: string]: ComponentType<*> },
  fallback?: ComponentType<*>,
}

const Fallback = () => null

export const renderFromObject = (
  tree: ConvertedElement,
  params?: RenderParams,
) => {
  const components = (params && params.components) || {}
  const fallback = (params && params.fallback) || Fallback

  const createChild = c => {
    if (c == null) {
      return null
    }
    if (Array.isArray(c)) {
      return c.map(createChild)
    }
    if (
      typeof c === 'boolean' ||
      typeof c === 'number' ||
      typeof c === 'string'
    ) {
      return c
    }
    if (c.type) {
      return createFromObject(c)
    }
    return null
  }

  const createFromObject = (converted: ConvertedElement) => {
    const component = components[converted.type] || fallback

    if (!converted.props) {
      return createElement(component)
    }

    const { children, ...props } = converted.props
    if (children != null) {
      props.children = createChild(children)
    }
    return createElement(component, props)
  }

  return createFromObject(tree)
}

type RendererProps = {
  json?: string,
  tree?: ConvertedElement,
} & RenderParams

export class Renderer extends PureComponent<RendererProps> {
  props: RendererProps

  render() {
    const { json, tree, ...params } = this.props
    let obj
    if (tree) obj = tree
    else if (json) obj = JSON.parse(json) // flowlint-line sketchy-null-string:off

    return obj ? renderFromObject(obj, params) : null
  }
}
