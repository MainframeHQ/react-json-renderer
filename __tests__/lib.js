/* global it */

import React, { Component, Fragment } from 'react'
import renderer from 'react-test-renderer'

import {
  convertToObject,
  convertToJSON,
  renderFromObject,
  Renderer,
} from '../src'

it('exposes the convertToObject, convertToJSON, renderFromObject and Renderer', () => {
  expect(convertToObject).toBeDefined()
  expect(convertToJSON).toBeDefined()
  expect(renderFromObject).toBeDefined()
  expect(Renderer).toBeDefined()
})

it('works end-to-end', () => {
  const Text = 'Text'
  const View = 'View'

  const Header = ({ title, ...props }) => (
    <View {...props} type="header">
      <Text>{title}</Text>
    </View>
  )
  const Wrapper = ({ children }) => <View type="wrapper">{children}</View>

  const json = convertToJSON(
    <Wrapper>
      <Header key="header" title="Title" />
      <View key="contents" type="main">
        <Text>Contents</Text>
      </View>
    </Wrapper>,
    {
      space: 2,
    },
  )

  const components = {
    Text: ({ children }) => <span className="text">{children}</span>,
    View: ({ children, type }) => (
      <div className={`view view-${type}`}>{children}</div>
    ),
  }

  const tree = renderer
    .create(<Renderer components={components} json={json} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Fragment works end-to-end', () => {
  const Text = 'Text'

  const json = convertToJSON(
    <Fragment>
      <Text key="1">hello</Text>
      <Text key="2">test</Text>
    </Fragment>,
    { space: 2 },
  )

  const components = {
    Text: ({ children }) => <span className="text">{children}</span>,
  }

  const tree = renderer
    .create(<Renderer components={components} json={json} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
