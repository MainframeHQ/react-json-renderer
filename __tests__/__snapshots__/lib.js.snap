// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fragment works end-to-end 1`] = `
Array [
  <span
    className="text"
  >
    hello
  </span>,
  <span
    className="text"
  >
    test
  </span>,
]
`;

exports[`works end-to-end 1`] = `
<div
  className="view view-wrapper"
>
  <div
    className="view view-header"
  >
    <span
      className="text"
    >
      Title
    </span>
  </div>
  <div
    className="view view-main"
  >
    <span
      className="text"
    >
      Contents
    </span>
  </div>
</div>
`;
